### PR TITLE
Tox, for parallel test goodness

### PR DESCRIPTION
**Pull request by efischer19**
# [EDU-768](https://openedx.atlassian.net/browse/EDUCATOR-768)

Attempts to split each of the following into separate test contexts on travis:
- django 1.8 python tests
- django 1.11 python tests
- quality checks
- js tests

### CODE REVIEW

**Comment by efischer19** [comment]
This has now morphed into a "modernize ORA" PR. With the goal of faster, smaller, parallelized tests, I'll be doing the following:

- [ ] modify installation, there's no reason we need to specify apt-packages. Requirements were also getting silly.
- [ ] remove AI grading, that eats a lot of testing cycles for a feature that isn't enabled anywhere. I'll double-check with product to ensure this is acceptable.
- [ ] tests will be more direct, and not make use of "travis calls tox calls makefile calls a bash script"

**Comment by efischer19** [comment]
Running into an order-dependent flaky test in test_training.py, under `--hashseed 2507972091`

File: requirements/tox.txt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+# Third Party Requirements
+boto>=2.32.1,<3.0.0
+python-swiftclient==3.1.0
+defusedxml==0.5.0
+django-model-utils==3.0.0
+django-nose==1.4.4
+git+https://github.com/edx/django-rest-framework.git@1ceda7c086fddffd1c440cc86856441bbf0bd9cb#egg=djangorestframework==3.6.3
+dogapi==1.11.1
+jsonfield==2.0.2
```
Comment (efischer19) [inline requirements/tox.txt]: @MichaelRoytman just saw your recent branch - I'm upgrading `jsonfield` here as part of https://openedx.atlassian.net/browse/EDUCATOR-768

File: requirements/tox.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# edX Internal Requirements
+git+https://github.com/edx/XBlock.git@efischer/play_nice_with_py27#egg=XBlock==1.0.0
```
Comment (efischer19) [inline requirements/tox.txt]: Relies on https://github.com/edx/XBlock/pull/365, which is currently unmerged.

**Comment by efischer19** [comment]
Ready for review! @cahrens has volunteered to be a reviewer (thanks again!); @edx/educator-dahlia FYI as well.

Other subtasks before I close EDUCATOR-409:

- [ ] upload to PyPI, I'll do this work in a separate PR for easier reviewing
- [ ] get https://github.com/edx/XBlock/pull/365 merged
- [ ] update Confluence docs for install/test/developing against edx-ora2

File: logs/README.txt [hunk 3]
```diff
@@ -1,7 +0,0 @@
-Log files:
-
-apps_info.log = INFO level logging for all edx-ora2 apps and OpenAssessmentBlock
```
Comment (cahrens) [inline logs/README.txt]: Should this information be captured someplace else? It seems like it could be useful (assuming it is still true).

File: openassessment/assessment/models/ai.py [hunk 4]
```diff
@@ -1,874 +1 @@
...
-            InvalidRubricSelection
-            DatabaseError
-
-        """
-        self.assessment = Assessment.create(
-            self.rubric, self.algorithm_id, self.submission_uuid, AI_ASSESSMENT_TYPE
-        )
-        AssessmentPart.create_from_option_points(self.assessment, criterion_scores)
-        self.mark_complete_and_save()
+# This file is empty, but we cannot delete it b/c historical migration records refer to it.
```
Comment (cahrens) [inline openassessment/assessment/models/ai.py]: Is this still true given our discussion?

File: openassessment/assessment/tasks.py [hunk 5]
```diff
@@ -1,7 +0,0 @@
-"""
-Celery looks for tasks in this module,
-so import the tasks we want the workers to implement.
-"""
-# pylint:disable=W0611
-from .worker.training import train_classifiers, reschedule_training_tasks
-from .worker.grading import grade_essay, reschedule_grading_tasks
```
Comment (cahrens) [inline openassessment/assessment/tasks.py]: Why is worker.grading removed?

File: openassessment/fileupload/backends/swift.py [hunk 6]
```diff
@@ -32,10 +35,11 @@ def get_upload_url(self, key, content_type):
         key, url = get_settings()
         try:
             temp_url = swiftclient.utils.generate_temp_url(
-                path='%s/%s/%s' % (url.path, bucket_name, key_name),
+                path='/v%s%s/%s/%s' % (SWIFT_BACKEND_VERSION, url.path, bucket_name, key_name),
```
Comment (cahrens) [inline openassessment/fileupload/backends/swift.py]: Will this cause any incompatibilities for people who are already using the Swift backend?

File: openassessment/management/commands/performance_test_for_submissions.py [hunk 7]
```diff
@@ -1,74 +0,0 @@
...
-"""
-import random
-import datetime
-
-from django.core.management.base import BaseCommand
-
-from openassessment.assessment.models import PeerWorkflow
-
-
-class Command(BaseCommand):
```
Comment (cahrens) [inline openassessment/management/commands/performance_test_for_submissions.py]: Why was this file deleted?

File: openassessment/management/tests/test_upload_oa_data.py [hunk 8]
```diff
@@ -69,4 +67,4 @@ def test_upload(self):
 
         # Expect that we generated a URL for the bucket
         url = cmd.history[0]['url']
-        self.assertIn("https://{}".format(self.BUCKET_NAME), url)
+        self.assertIn("ttps://s3.amazonaws.com/{}".format(self.BUCKET_NAME), url)
```
Comment (cahrens) [inline openassessment/management/tests/test_upload_oa_data.py]: Why is the "h" missing from the front?

File: openassessment/tests/test_data.py [hunk 9]
```diff
@@ -363,13 +363,15 @@ class TestOraAggregateDataIntegration(TransactionCacheResetTest):
...
         super(TestOraAggregateDataIntegration, self).setUp()
+        self.maxDiff = None
         # Create submissions and assessments
         self.submission = self._create_submission(STUDENT_ITEM)
         self.scorer_submission = self._create_submission(SCORER_ITEM)
         self.earned_points = 1
         self.possible_points = 2
         peer_api.get_submission_to_assess(self.scorer_submission['uuid'], 1)
         self.assessment = self._create_assessment(self.scorer_submission['uuid'])
+        self.assertEqual(self.assessment['parts'][0]['criterion']['label'], "criterion_1")
```
Comment (cahrens) [inline openassessment/tests/test_data.py]: What caused this change (and the lines below, 475 and following?

File: openassessment/xblock/static/sass/oa/views/_oa-base.scss [hunk 10]
```diff
@@ -1086,7 +1086,7 @@
     }
 
     .action--save {
-      @extend %action-button
+      @extend %action-button !optional
```
Comment (cahrens) [inline openassessment/xblock/static/sass/oa/views/_oa-base.scss]: Why was this necessary?

File: openassessment/xblock/static/sass/oa/views/_oa-base.scss [hunk 11]
```diff
@@ -1086,7 +1086,7 @@
...
     }
 
     .feedback {
```
Comment (cahrens) [inline openassessment/xblock/static/sass/oa/views/_oa-base.scss]: Is there CSS that should be deleted related to AI grading? I'm thinking of the Studio editors, as well as the LMS view.

File: openassessment/xblock/studio_mixin.py [hunk 12]
```diff
@@ -379,19 +347,13 @@ def _editor_assessments_order_context(self):
         # since the user last saved their ordering.
         effective_order = copy.deepcopy(DEFAULT_EDITOR_ASSESSMENTS_ORDER)
 
-        # If the problem already contains example-based assessment
-        # then allow the editor to display example-based assessments,
-        # which is not included in the default
-        enabled_assessments = [asmnt['name'] for asmnt in self.valid_assessments]
```
Comment (cahrens) [inline openassessment/xblock/studio_mixin.py]: Have you checked that no ORA instances in prod/Edge have AI steps defined? Just wondering how they would fail.

File: scripts/run-pep8.sh [hunk 13]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-MAX_PEP8_VIOLATIONS=106
+MAX_PEP8_VIOLATIONS=62  # TODO nomerge - but also stop failing during development
```
Comment (cahrens) [inline scripts/run-pep8.sh]: Don't forget to follow up on this.

File: scripts/run-pylint.sh [hunk 14]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-MAX_PYLINT_VIOLATIONS=438
+MAX_PYLINT_VIOLATIONS=495  # TODO nomerge - but also stop failing during development
```
Comment (cahrens) [inline scripts/run-pylint.sh]: Don't forget to follow up on this.

File: test/performance/README.rst [hunk 15]
```diff
@@ -1,50 +0,0 @@
-Performance Tests
-=================
-
-1. Install performance test requirements:
```
Comment (cahrens) [inline test/performance/README.rst]: No more performance tests?

**Comment by cahrens** [comment]
What are the .gitkeep files?

**Comment by cahrens** [comment]
What is the situation with the course.tar.gz file (I don't understand the status from the files tab)?

**Comment by cahrens** [comment]
@efischer19 I'm done slogging through this-- wow, a lot of great cleanup!

File: logs/README.txt (hunk 3 above)
Comment (efischer19) [inline logs/README.txt]: I don't believe this info is true any longer, edx-platform defines logging locations (which work with splunk), and this particular file is >3 years past its last update.

File: openassessment/assessment/models/ai.py (hunk 4 above)
Comment (efischer19) [inline openassessment/assessment/models/ai.py]: I will double check on this one - it could be in a weird intermediary state from my internal debate about dealing with migrations.

File: openassessment/assessment/tasks.py (hunk 5 above)
Comment (efischer19) [inline openassessment/assessment/tasks.py]: Everything under worker has been removed - the idea was to "train" your AI classifiers with a given data set ahead of time, then "grade" new essays as a separate phase. We don't need any references to either of these operations.

File: openassessment/fileupload/backends/swift.py (hunk 6 above)
Comment (efischer19) [inline openassessment/fileupload/backends/swift.py]: It shouldn't, this only affects newly generated upload urls.

File: openassessment/management/commands/performance_test_for_submissions.py (hunk 7 above)
Comment (efischer19) [inline openassessment/management/commands/performance_test_for_submissions.py]: I yanked out all performance tests as well, as they also heavily blurred the line between "python library" and "suite of scripts and programs to install, some of which may be python". The tests that were here were incredibly old, I think if we need performance testing against ora we're better off rewriting it in https://github.com/edx/edx-load-tests using the modern locust setup described there.

File: openassessment/management/tests/test_upload_oa_data.py (hunk 8 above)
Comment (efischer19) [inline openassessment/management/tests/test_upload_oa_data.py]: Terrible copy-paste job by the dev who put this here, what a bum that guy must be (thanks, will fix)

File: openassessment/tests/test_data.py (hunk 9 above)
Comment (efischer19) [inline openassessment/tests/test_data.py]: It was a flaky test issue. A particular hashseed value caused the assessment to store out-of-order values, until I made [this fix](https://github.com/edx/edx-ora2/pull/1023/files/5d5921287a9c00b3935b86756cd960f64374b678#diff-c09750592365024b37e7601ab2b68bad). Now, this test ensures that `self.assessment['parts'][0]` matches up to the first criterion and appears first in the downloaded data.

File: openassessment/xblock/static/sass/oa/views/_oa-base.scss (hunk 10 above)
Comment (efischer19) [inline openassessment/xblock/static/sass/oa/views/_oa-base.scss]: It was failing quality builds, not sure why it hadn't been before. I'll double-check to ensure that's still the case.

File: openassessment/xblock/static/sass/oa/views/_oa-base.scss (hunk 11 above)
Comment (efischer19) [inline openassessment/xblock/static/sass/oa/views/_oa-base.scss]: Oh, good call. I'll do an audit of that before the next round of review

File: openassessment/xblock/studio_mixin.py (hunk 12 above)
Comment (efischer19) [inline openassessment/xblock/studio_mixin.py]: From my database poking, I don't believe there are any. I'll verify that, and set up such a problem locally to ensure we can handle that situation correctly.

File: test/performance/README.rst (hunk 15 above)
Comment (efischer19) [inline test/performance/README.rst]: Correct, I've eliminated them from this repo for reasons outlined above.

**Comment by efischer19** [comment]
`.gitkeep` files [are placed in a directory to ensure it exists on checkout](https://stackoverflow.com/questions/7229885/what-are-the-differences-between-gitignore-and-gitkeep). I thought those directories were overkill and have removed them, so the files are deleted.

course.tar.gz is a binary file that had been included for performance testing. I think storing it in git is the wrong move, but I also just realized that removing it may cause problems with https://tools-edx-jenkins.edx.org/job/ora2/job/add-course-to-sandbox/. I'll add that to my list of things to look into.

Thanks for the review! I'll let you know when I've addressed everything.

**Comment by cahrens** [comment]
@efischer19 related to the .gitkeep files, I was confused by this "grey" state. What does it mean?

image

**Comment by efischer19** [comment]
Oh! I think that's github's UI having a hard time with "This file of size 0 has been deleted" - it does correctly show that no lines have changed, but https://github.com/edx/edx-ora2/blob/5d5921287a9c00b3935b86756cd960f64374b678/storage/dev/.gitkeep 404s, so there's no file there.

File: requirements/tox.txt (hunk 2 above)
Comment (efischer19) [inline requirements/tox.txt]: Merged and updated

File: openassessment/xblock/static/sass/oa/views/_oa-base.scss (hunk 10 above)
Comment (efischer19) [inline openassessment/xblock/static/sass/oa/views/_oa-base.scss]: Example failure: https://travis-ci.org/edx/edx-ora2/jobs/257446578

File: openassessment/xblock/static/sass/oa/views/_oa-base.scss (hunk 11 above)
Comment (efischer19) [inline openassessment/xblock/static/sass/oa/views/_oa-base.scss]: I've poked around - no files are obviously AI grading exclusive, and I can't see any obvious removals in the existing files. Judging by that and the fact that very little HTML needed to be removed, I think we're safe assuming that this feature never had a lot of dedicated front-end customization done.

**Comment by efischer19** [comment]
The course that gets added to sandboxes for acceptance testing lives under `scripts/data`, the one I've deleted was under `test/performance`, so [jenkins can still do its job](https://tools-edx-jenkins.edx.org/job/ora2/job/add-course-to-sandbox/132)

**Comment by efischer19** [comment]
I've confirmed (locally) that a problem with a defined example-based assessment step will still render properly.

**Comment by efischer19** [comment]
Alright @cahrens, I believe this is ready again, I've addressed all your comments from before. The diff of the latest commit looks scary, but a large portion of it was me running `isort` to standardize all our import statements (which fixed a huge number of quality errors).

I'm also running acceptance tests at https://gocd.tools.edx.org/go/pipelines/value_stream_map/build_ora2_sandbox/142.

File: requirements/base.txt [hunk 16]
```diff
@@ -1,31 +1,23 @@
-# Include the requirements we're caching as "wheel" archives
-# to speed up the test builds.
--r wheels.txt
-
 # edX Internal Requirements
-git+https://github.com/edx/XBlock.git@xblock-0.4.12#egg=XBlock==0.4.12
+# edx-submissions>=2.0.6,<3.0.0 #TODO: update after 2.0.6 release
+git+https://github.com/edx/edx-submissions.git@efischer/loosen_reqs#egg=edx-submissions==2.0.6 # TODO: and remove this line
```
Comment (efischer19) [inline requirements/base.txt]: These two lines are effectively "I'm waiting for https://github.com/edx/edx-submissions/pull/65 to merge before this"

File: Makefile [hunk 17]
```diff
@@ -2,6 +2,7 @@
 # Install commands
 ##################
 install-python:
+	pip install -r requirements/django.txt
```
Comment (cahrens) [inline Makefile]: Why a separate file?

File: openassessment/workflow/migrations/0002_auto_20170724_1537.py [hunk 18]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+import uuid
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
```
Comment (cahrens) [inline openassessment/workflow/migrations/0002_auto_20170724_1537.py]: What caused this migration?

File: requirements/base.txt [hunk 19]
```diff
@@ -1,3 +1,23 @@
-django>=1.8,<2.0
+# edX Internal Requirements
+# edx-submissions>=2.0.6,<3.0.0 #TODO: update after 2.0.6 release
+git+https://github.com/edx/edx-submissions.git@efischer/loosen_reqs#egg=edx-submissions==2.0.6 # TODO: and remove this line
+git+https://github.com/edx/django-rest-framework.git@1ceda7c086fddffd1c440cc86856441bbf0bd9cb#egg=djangorestframework==3.6.3
```
Comment (cahrens) [inline requirements/base.txt]: Why a specific commit of django-rest-framework?

File: setup.cfg [hunk 20]
```diff
@@ -0,0 +1,9 @@
+[isort]
+line_length = 120
+known_third_party = ddt,mock
+known_django = django
+known_djangoapp = jsonfield,model_utils
+known_edx =
```
Comment (cahrens) [inline setup.cfg]: known_edx?

File: Makefile (hunk 17 above)
Comment (cahrens) [inline Makefile]: I see-- because of the tox configuration. Never mind.

File: settings/base.py [hunk 21]
```diff
@@ -93,6 +92,12 @@
     'django.template.loaders.filesystem.Loader',
```
Comment (cahrens) [inline settings/base.py]: These are the default values, so TEMPLATE_LOADERS can be removed.

https://docs.djangoproject.com/en/1.11/ref/templates/upgrading/

File: settings/base.py [hunk 22]
```diff
@@ -3,7 +3,6 @@
 """
 
 import os
-from celery import Celery
 
 DEBUG = True
 TEMPLATE_DEBUG = DEBUG
```
Comment (cahrens) [inline settings/base.py]: I believe you should remove TEMPLATE_DEBUG. You can include 'debug' in the TEMPLATES options.

For example,

````
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [
            os_path.join(PROJECT_PATH, 'templates'),
        ],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                "django.contrib.auth.context_processors.auth",
                "django.template.context_processors.debug",
                "django.template.context_processors.i18n",
                "django.template.context_processors.media",
                "django.template.context_processors.static",
                "django.template.context_processors.request",
                "django.template.context_processors.tz",
                "django.contrib.messages.context_processors.messages",
                "sekizai.context_processors.sekizai",
            ],
            'debug': DEBUG,
        },
    },
]

Comment (cahrens) [inline settings/base.py]: Also, STATICFILES_DIRS and STATICFILES_FINDERS can be removed.

https://docs.djangoproject.com/en/1.11/ref/settings/#std:setting-STATICFILES_FINDERS

**Comment by cahrens** [comment]
This seems very odd:

image

**Comment by cahrens** [comment]
@efischer19 done re-reviewing-- just a few more comments/questions.

File: openassessment/workflow/migrations/0002_auto_20170724_1537.py (hunk 18 above)
Comment (efischer19) [inline openassessment/workflow/migrations/0002_auto_20170724_1537.py]: It's the same thing that's happening [here](https://github.com/edx/edx-submissions/pull/65/files#r130444360) - We can remove the `django_extensions` dependency fairly easily, but that touches a field on a model, which leads to a migration. Said migration *should* be a no-op, but there are some pending operational questions to sort out.

The migration should have been in the last commit; I had forgotten that it needed to be generated and ran into a "please run makemigrations" warning while fixing other things. I'll work out the details with Kevin on my other PR and update this one with the results of that conversation.

File: requirements/base.txt (hunk 19 above)
Comment (efischer19) [inline requirements/base.txt]: @doctoryes set this requirement to a specific commit in https://github.com/edx/edx-ora2/pull/1024, I'm just moving it to another file

File: setup.cfg (hunk 20 above)
Comment (efischer19) [inline setup.cfg]: Yeah, it has something to do with isort configuration. I copied this over from https://github.com/edx/edx-celeryutils/blob/master/setup.cfg, I think we can remove the `known_edx` line

**Comment by efischer19** [comment]
Thanks for the review @cahrens! I think I've answered all your questions, will get this updated witch code changes and ready for merge in the next day or so.

File: openassessment/workflow/migrations/0002_auto_20170724_1537.py (hunk 18 above)
Comment (efischer19) [inline openassessment/workflow/migrations/0002_auto_20170724_1537.py]: Update: we're pushing the submissions upgrade through independently, as it had an identical migration. Expectation is that it runs as a no-op, we're going to verify that as submissions goes out since it's a smaller table.

**Comment by efischer19** [comment]
We're back - latest commit has:

- updated location of developer wiki links (I'm going to be editing those soon to explain docker devstack setup FYI)
- finalized edx-submissions requirement to 2.0.6, since that merged
- lowered pylint error limit, since I was well below it
- cleaned up settings/base.py as suggested
- preemptively fixed a flaky test I noticed earlier this morning

Once tests have passed and everyone is happy, I'll get this squashed. merged, and deployed

**Comment by cahrens** [comment]
👍

**Comment by efischer19** [comment]
🎆 

Thanks for the help here @cahrens!

I'm going to get this deployed tomorrow

**Comment by cahrens** [comment]
🚀
